Include/cpython/object.h

```diff
@@ -4,6 +4,9 @@
 
 PyAPI_FUNC(void) _Py_NewReference(PyObject *op);
 
+PyAPI_FUNC(PyObject *) _PyType_DisabledNew(PyTypeObject *, PyObject *,
```

Comment:
Can we move this to the inernal API maybe? @vstinner what do you think?

Reply:
Wouldn't this be useful for third party C API extensions?

Reply:
It could, but the less exposed C-API the better. Exposing too much is how we ended with hundreds of functions that are super difficult to deprecate or change. 

My thoughts are: unless is super obvious that should be public, the fewer the better.

Reply:
I agree. Let's start in the internal API. If at a later point, we're unhappy with this solution, it won't hurt to tear it out.

Reply:
So, this means a lot of the extension modules will have to be built with `Py_BUILD_CORE_BUILTIN` defined. I think this PR has become too large and unwieldy. I suggest adding `_PyType_DisabledNew` in one PR, and then add the regression fixes in two or three batches.

Reply:
@pablogsal, would you mind taking a look at commit f74d5749b44d422880b12f02eb93279ca0d27c54. I had two use `PyAPI_FUNC` in order to build, but I don't understand why; `extern` should have been enough. I'm don't know the build system well enough to sort it out. Perhaps you can shed some light onto this?

Reply:
> had two use `PyAPI_FUNC` in order to build, but I don't understand why;

That's because we are hiding symbols by default and `PyAPI_FUNC` marks them as exported. You can use extern if the definition is in a compilation unit that will be built into the core, though.

Reply:
Yes, and `Objects/typeobject.c` is built into the core AFAICS, so it should be fine. But I'm getting symbol not found errors when I try to declare it just using `extern`.

Reply:
Uncommenting the affected modules in `Modules/Setup` solves it. But that feels like the wrong fix.

Reply:
I can take a look today at night, but otherwise mark it as exported by the API as is now and we can fix it later.

Reply:
Ok, thanks, Pablo! I'll leave the PR as it is for now.

Reply:
Another option is to revert f74d5749b44d422880b12f02eb93279ca0d27c54 and leave it in Include/cpython. I let you and Victor decide.